pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/retention/RetentionManager.java

```diff
@@ -199,6 +225,70 @@ private void manageRetentionForRealtimeTable(String realtimeTableName, Retention
     }
   }
 
+  @VisibleForTesting
+  void manageRetentionForHybridTable(TableConfig realtimeTableConfig, TableConfig offlineTableConfig) {
+    LOGGER.info("Managing retention for hybrid table: {}", realtimeTableConfig.getTableName());
+    List<String> segmentsToDelete = new ArrayList<>();
+    String realtimeTableName = realtimeTableConfig.getTableName();
+    String rawTableName = TableNameBuilder.extractRawTableName(realtimeTableName);
+    String offlineTableName = TableNameBuilder.OFFLINE.tableNameWithType(rawTableName);
+    try {
+      ZkHelixPropertyStore<ZNRecord> propertyStore = _pinotHelixResourceManager.getPropertyStore();
+      Schema schema = ZKMetadataProvider.getTableSchema(propertyStore, offlineTableName);
+      if (schema == null) {
+        throw new RuntimeException("Failed to get schema for table: " + offlineTableName);
```

Comment:
(minor) We usually throw `IllegalStateException`. You may use `Preconditions.checkState()` for concise. Same for other places